storage/index/iterator.go

```diff
@@ -0,0 +1,131 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/m3db/m3db/x/xpool"
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+)
+
+var (
+	errInvalidResultMissingNamespace = errors.New("corrupt data, unable to extract namespace")
+)
+
+type idsIter struct {
+	nsID        ident.ID
+	iter        segment.ResultsIter
+	err         error
+	idPool      ident.Pool
+	wrapperPool xpool.CheckedBytesWrapperPool
+
+	currentID   ident.ID
+	currentTags ident.Tags
+}
+
+// NewIterator returns a new Iterator backed by
+// a segment.ResultsIter.
+func NewIterator(
+	nsID ident.ID,
+	iter segment.ResultsIter,
+	opts Options,
+) Iterator {
+	i := &idsIter{
+		nsID:        nsID,
+		iter:        iter,
+		idPool:      opts.IdentifierPool(),
+		wrapperPool: opts.CheckedBytesWrapperPool(),
+	}
+	return i
+}
+
+func (i *idsIter) Next() bool {
+	// release any held resources
+	i.release()
+
+	if i.err != nil {
+		return false
+	}
+
+	// check underlying iterator
+	next := i.iter.Next()
+	if !next {
+		i.err = i.iter.Err()
+		return next
+	}
+
+	d, _ := i.iter.Current()
+	err := i.parseAndStore(d)
+	if err != nil {
+		i.err = err
+		return false
+	}
+
+	return true
+}
+
+func (i *idsIter) Current() (namespaceID ident.ID, seriesID ident.ID, tags ident.Tags) {
+	return i.nsID, i.currentID, i.currentTags
+}
+
+func (i *idsIter) Err() error {
+	return i.err
+}
+
+func (i *idsIter) release() {
+	if i.currentID != nil {
+		i.currentID.Finalize()
+		i.currentID = nil
+	}
+	if i.currentTags != nil {
+		i.currentTags.Finalize()
+		i.currentTags = nil
+	}
+	// NB(prateek): iterator doesn't own nsID so we don't finalize it here.
+}
+
+func (i *idsIter) wrapBytes(bytes []byte) checked.Bytes {
+	return i.wrapperPool.Get(bytes)
+}
+
+func (i *idsIter) parseAndStore(d doc.Document) error {
+	idFound := false
+	for _, f := range d.Fields {
+		if !idFound && bytes.Equal(f.Name, ReservedFieldNameID) {
+			i.currentID = i.idPool.BinaryID(i.wrapBytes(f.Value))
```

Comment:
aside: This whole wrap bytes thing makes me think maybe we should've been using `checked.Bytes` in the m3ninx code, but that's neither here nor there now.

Reply:
Not so sure about that. If we're going to have external consumers of `m3ninx`, the `[]byte` interface is a lot more familiar to Go users than `checked.Bytes` would be. 

Can chat more offline.